### PR TITLE
Query ip

### PR DESCRIPTION
**Pull request by dannybackx**
Minimal testing performed :-)

This goes with a PR for el-client that I need to create too.

Functionality provided with this PR is best summarized by the additional codes in cmd/cmd.h : 
+  CMD_GET_WIFI_INFO,   // query ip address info
+  CMD_SET_WIFI_INFO,   // set ip address info
+  CMD_WIFI_GET_APCOUNT = 50,    // Get number of access points
+  CMD_WIFI_GET_APNAME,     // Get the name (ssid) of an access point
+  CMD_WIFI_SELECT_SSID,     // select a network to connect to
+  CMD_WIFI_SIGNAL_STRENGTH,    // Query RSSI (wifi signal strength)
+  CMD_WIFI_GET_SSID,           // Query SSID/network name currently connected to
+  CMD_WIFI_START_SCAN,         // Trigger a scan (takes a long time)

### CODE REVIEW

File: cmd/handlers.c [hunk 1]
```diff
@@ -197,3 +245,114 @@ cmdAddCallback(CmdPacket *cmd) {
 
   cmdAddCb(name, cmd->value); // save the sensor callback
 }
+
+// Query the number of wifi access points
+static void ICACHE_FLASH_ATTR cmdWifiGetApCount(CmdPacket *cmd) {
+  int n = wifiGetApCount();
+  os_printf("WifiGetApCount : %d\n", n);
```
Comment (tve) [inline cmd/handlers.c]: The `os_printf` in this file need to be converted to `DBG` so they can be disabled. See 8 lines up, for example.

File: cmd/handlers.c [hunk 2]
```diff
@@ -197,3 +245,114 @@ cmdAddCallback(CmdPacket *cmd) {
...
+  if (argc != 2)
+    return;
+
+  int len = cmdArgLen(&req);
+  if (len == 1) {
+    // Assume this is the index
+    uint8_t ix;
+    cmdPopArg(&req, &ix, 1);
+    len = cmdArgLen(&req);
+    pass = (char *)os_malloc(len+2);
```
Comment (tve) [inline cmd/handlers.c]: Shouldn't that be len+1 ?

File: cmd/handlers.c [hunk 3]
```diff
@@ -197,3 +245,114 @@ cmdAddCallback(CmdPacket *cmd) {
...
+    os_printf("SelectSSID(%s,%s)", ssid, pass);
+    connectToNetwork(ssid, pass);
+  }
+}
+
+/*
+ * Once we're attached to some wireless network, choose not to pick up address from
+ * DHCP or so but set our own.
+ */
+static void ICACHE_FLASH_ATTR cmdSetWifiInfo(CmdPacket *cmd) {
```
Comment (tve) [inline cmd/handlers.c:339]: unimplemented? then remove for now?

File: esp-link/cgiservices.c [hunk 4]
```diff
@@ -1,4 +1,5 @@
 #include <esp8266.h>
+#include "cmd.h"
```
Comment (tve) [inline esp-link/cgiservices.c]: why is this needed?

File: esp-link/cgiwifi.c [hunk 5]
```diff
@@ -238,6 +241,16 @@ static int ICACHE_FLASH_ATTR cgiWiFiStartScan(HttpdConnData *connData) {
   return HTTPD_CGI_DONE;
 }
 
+// Start scanning, API interface
```
Comment (tve) [inline esp-link/cgiwifi.c]: Can you consolidate the code a bit and have one function call the other? Flash isn't infinite...

File: esp-link/cgiwifi.c [hunk 6]
```diff
@@ -945,3 +964,58 @@ void ICACHE_FLASH_ATTR wifiInit() {
...
+ICACHE_FLASH_ATTR void wifiGetApName(int i, char *ptr) {
+  if (i < 0)
+    return;
+  if (i >= cgiWifiAps.noAps)
+    return;
+
+  if (ptr != 0)
+    strncpy(ptr, cgiWifiAps.apData[i]->ssid, 32);
+
+  os_printf("AP %s\n", cgiWifiAps.apData[i]->ssid);
```
Comment (tve) [inline esp-link/cgiwifi.c]: DBG or delete...

File: esp-link/cgiwifi.c [hunk 7]
```diff
@@ -945,3 +964,58 @@ void ICACHE_FLASH_ATTR wifiInit() {
...
+}
+
+// This may not belong here : called from cmd/handlers.c
+// But it's good to have similar functionality close to each other.
+// This performs functions similar to cgiWiFiConnect()
+int ICACHE_FLASH_ATTR wifiConnect(char *ssid, char *pass) {
+// Danny
```
Comment (tve) [inline esp-link/cgiwifi.c]: unimplemented -> remove for now?

File: esp-link/cgiwifi.c [hunk 8]
```diff
@@ -945,3 +964,58 @@ void ICACHE_FLASH_ATTR wifiInit() {
...
 }
+
+// Access functions for cgiWifiAps
+int ICACHE_FLASH_ATTR wifiGetApCount() {
+  if (cgiWifiAps.scanInProgress)
+    return 0;
+  return cgiWifiAps.noAps;
+}
+
+ICACHE_FLASH_ATTR void wifiGetApName(int i, char *ptr) {
```
Comment (tve) [inline esp-link/cgiwifi.c]: Add a comment for the function?

File: esp-link/cgiwifi.c [hunk 9]
```diff
@@ -945,3 +964,58 @@ void ICACHE_FLASH_ATTR wifiInit() {
...
+
+// This may not belong here : called from cmd/handlers.c
+// But it's good to have similar functionality close to each other.
+// This performs functions similar to cgiWiFiConnect()
+int ICACHE_FLASH_ATTR wifiConnect(char *ssid, char *pass) {
+// Danny
+  return 0;
+}
+
+ICACHE_FLASH_ATTR int wifiSignalStrength(int i) {
```
Comment (tve) [inline esp-link/cgiwifi.c]: Add a comment for the function?

File: esp-link/cgiwifi.c [hunk 10]
```diff
@@ -945,3 +964,58 @@ void ICACHE_FLASH_ATTR wifiInit() {
...
+    rssi = wifi_station_get_rssi();	// Current network's signal strength
+  else if (i >= cgiWifiAps.noAps)
+    rssi = 0;				// FIX ME
+  else
+    rssi = cgiWifiAps.apData[i]->rssi;	// Signal strength of any known network
+
+  return rssi;
+}
+
+void ICACHE_FLASH_ATTR cmdWifiQuerySSID(CmdPacket *cmd) {
```
Comment (tve) [inline esp-link/cgiwifi.c]: This should be in handlers.c not here...

File: esp-link/cgiwifi.c [hunk 11]
```diff
@@ -945,3 +964,58 @@ void ICACHE_FLASH_ATTR wifiInit() {
...
+  CmdRequest req;
+  cmdRequest(&req, cmd);
+  uint32_t callback = req.cmd->value;
+
+  struct station_config conf;
+  // bool res = wifi_station_get_config(&conf);
```
Comment (tve) [inline esp-link/cgiwifi.c]: ???

File: cmd/handlers.c [hunk 12]
```diff
@@ -197,3 +245,114 @@ cmdAddCallback(CmdPacket *cmd) {
...
+  os_printf("SetWifiInfo()\n");
+}
+
+static void ICACHE_FLASH_ATTR cmdWifiSignalStrength(CmdPacket *cmd) {
```
Comment (tve) [inline cmd/handlers.c]: Add a comment for the function?

File: esp-link/status.c [hunk 13]
```diff
@@ -3,6 +3,7 @@
 #include <esp8266.h>
 #include "config.h"
 #include "serled.h"
+#include "cmd.h"
```
Comment (tve) [inline esp-link/status.c]: Why is this needed?

File: esp-link/mqtt_client.c [hunk 14]
```diff
@@ -114,4 +116,23 @@ mqtt_client_on_data(MqttDataCallback dataCb) {
   data_cb = dataCb;
 }
 
+// Command handler for MQTT information
+void ICACHE_FLASH_ATTR cmdMqttGetClientId(CmdPacket *cmd) {
```
Comment (tve) [inline esp-link/mqtt_client.c]: Should be in handlers.c?

File: esp-link/mqtt_client.c [hunk 15]
```diff
@@ -1,9 +1,11 @@
 #ifdef MQTT
 #include <esp8266.h>
+#include "cmd/cmd.h"
```
Comment (tve) [inline esp-link/mqtt_client.c]: This should be included here.

File: esp-link/main.c [hunk 16]
```diff
@@ -13,6 +13,7 @@
 #include "httpd.h"
 #include "httpdespfs.h"
 #include "cgi.h"
+#include "cmd.h"
```
Comment (tve) [inline esp-link/main.c]: This shouldn't be needed here?

File: esp-link/cgiwifi.c [hunk 17]
```diff
@@ -13,7 +13,9 @@ Cgi/template routines for the /wifi url.
  * ----------------------------------------------------------------------------
  */
 
+
 #include <esp8266.h>
+#include "cmd.h"
```
Comment (tve) [inline esp-link/cgiwifi.c]: This shouldn't be included here...

**Review by tve — CHANGES_REQUESTED** [review]
Looking nice overall! I made a slew of minor comments.

**Comment by dannybackx** [comment]
On 05/31/2017 06:50 AM, Thorsten von Eicken wrote:
> In esp-link/cgiservices.c <https://github.com/jeelabs/esp-link/pull/289#discussion_r119269617>:
> > @@ -1,4 +1,5 @@
>   #include <esp8266.h>
> +#include "cmd.h"
>
> why is this needed?
>

If I remove it, I get
 > CC esp-link/cgiservices.c
 > In file included from esp-link/cgiservices.c:2:0:
 > esp-link/cgiwifi.h:31:23: error: unknown type name 'CmdPacket'
 >  void cmdWifiQuerySSID(CmdPacket *cmd);
 >                        ^
 > esp-link/cgiwifi.h:32:23: error: unknown type name 'CmdPacket'
 >  void cmdWifiStartScan(CmdPacket *cmd);
 >                        ^
 > Makefile:521: recipe for target 'build/esp-link/cgiservices.o' failed

Or is this an error in my build environment ?

     Danny


-- 
Danny Backx - danny.backx@scarlet.be - http://danny.backx.info

File: cmd/handlers.c [hunk 18]
```diff
@@ -197,3 +250,162 @@ cmdAddCallback(CmdPacket *cmd) {
...
+  if (argc != 2)
+    return;
+
+  int len = cmdArgLen(&req);
+  if (len == 1) {
+    // Assume this is the index
+    uint8_t ix;
+    cmdPopArg(&req, &ix, 1);
+    len = cmdArgLen(&req);
+    pass = (char *)os_malloc(len+1);
```
Comment (tve) [inline cmd/handlers.c:309]: where is this ever freed? it looks like a heap allocation is totally unnecessary...

File: cmd/handlers.c [hunk 19]
```diff
@@ -197,3 +250,162 @@ cmdAddCallback(CmdPacket *cmd) {
...
+    wifiGetApName(ix, myssid);
+    myssid[32] = '\0';
+    connectToNetwork(myssid, pass);
+  } else {
+    ssid = os_malloc(len+1);
+    cmdPopArg(&req, ssid, len);
+    ssid[len] = 0;
+
+    len = cmdArgLen(&req);
+    pass = (char *)os_malloc(len+2);
```
Comment (tve) [inline cmd/handlers.c:325]: where is this ever freed?

File: cmd/handlers.c [hunk 20]
```diff
@@ -197,3 +250,162 @@ cmdAddCallback(CmdPacket *cmd) {
...
+    pass[len] = 0;
+
+    DBG("SelectSSID(%d,%s)", ix, pass);
+
+    char myssid[33];
+    wifiGetApName(ix, myssid);
+    myssid[32] = '\0';
+    connectToNetwork(myssid, pass);
+  } else {
+    ssid = os_malloc(len+1);
```
Comment (tve) [inline cmd/handlers.c:320]: where is this ever freed?

**Comment by tve** [comment]
Thanks much!

File: cmd/handlers.c (hunk 18 above)
Comment (dannybackx) [reply cmd/handlers.c:309]: Just committed a fix

File: cmd/handlers.c (hunk 20 above)
Comment (dannybackx) [reply cmd/handlers.c:320]: Just committed a fix

File: cmd/handlers.c (hunk 18 above)
Comment (dannybackx) [reply cmd/handlers.c:309]: Just committed a fix

Comment (tve) [reply cmd/handlers.c:309]: oops, sorry, after writing the comments I decided to merge your commit and fix the issues myself. It's already all on master :-) (No guarantees that it builds or works, I had to go to bed...)

Comment (dannybackx) [reply cmd/handlers.c:309]: No problem, I only spent a few minutes on it. Short enough not to notice that you'd already done it.
I'll check the build.

**Pull request merged**